File: apps/mobibix-android/app/src/main/java/com/aiyal/mobibix/data/network/KnowledgeApi.kt
package com.aiyal.mobibix.data.network

import retrofit2.http.*

// ─── Models ──────────────────────────────────────────────────────────────────

data class FaultType(val id: String, val name: String)

data class FaultDiagnosisStep(val id: String, val order: Int, val stepText: String)

data class FaultDiagnosis(
    val id: String,
    val faultTypeId: String,
    val description: String? = null,
    val steps: List<FaultDiagnosisStep> = emptyList(),
    val faultType: FaultType? = null
)

data class RepairNote(
    val id: String,
    val faultTypeId: String,
    val content: String,
    val videoUrl: String? = null,
    val status: String = "APPROVED",
    val source: String = "SYSTEM",
    val helpfulCount: Int = 0,
    val notHelpfulCount: Int = 0,
    val createdAt: String? = null
)

data class KnowledgeForJobResponse(
    val checklist: FaultDiagnosis? = null,
    val notes: List<RepairNote> = emptyList(),
    val suggestedFaultTypes: List<FaultType>? = null,
    val jobDetails: JobKnowledgeDetails? = null
)

data class JobKnowledgeDetails(
    val brand: String,
    val model: String,
    val problem: String,
    val phoneModelId: String? = null,
    val faultTypeId: String? = null
)

data class VoteNoteDto(val vote: String) // "helpful" or "notHelpful"

data class SubmitRepairNoteDto(
    val faultTypeId: String,
    val content: String,
    val videoUrl: String? = null,
    val phoneModelId: String? = null
)

// ─── API ─────────────────────────────────────────────────────────────────────

interface KnowledgeApi {

    @GET("api/mobileshop/knowledge/job/{jobCardId}")
    suspend fun getKnowledgeForJob(@Path("jobCardId") jobCardId: String): KnowledgeForJobResponse

    @GET("api/mobileshop/knowledge/fault-types")
    suspend fun listFaultTypes(): List<FaultType>

    @GET("api/mobileshop/knowledge/notes")
    suspend fun getRepairNotes(
        @Query("phoneModelId") phoneModelId: String? = null,
        @Query("faultTypeId") faultTypeId: String? = null
    ): List<RepairNote>

    @POST("api/mobileshop/knowledge/notes/{noteId}/vote")
    suspend fun voteOnNote(
        @Path("noteId") noteId: String,
        @Body dto: VoteNoteDto
    )

    @POST("api/mobileshop/knowledge/notes")
    suspend fun submitRepairNote(@Body dto: SubmitRepairNoteDto): RepairNote

    @GET("api/mobileshop/knowledge/checklist/{faultTypeId}")
    suspend fun getChecklist(@Path("faultTypeId") faultTypeId: String): FaultDiagnosis?
}

File: apps/mobibix-web/src/services/knowledge.api.ts
import { authenticatedFetch, extractData } from "./auth.api";

export interface FaultType {
  id: string;
  name: string;
}

export interface FaultDiagnosisStep {
  id: string;
  order: number;
  stepText: string;
}

export interface FaultDiagnosis {
  id: string;
  faultTypeId: string;
  tenantId?: string | null;
  description?: string | null;
  steps: FaultDiagnosisStep[];
  faultType: FaultType;
}

export type RepairKnowledgeStatus = "PENDING" | "APPROVED" | "REJECTED";
export type RepairKnowledgeSource = "SYSTEM" | "ADMIN" | "COMMUNITY";

export interface RepairNote {
  id: string;
  phoneModelId?: string | null;
  faultTypeId: string;
  content: string;
  videoUrl?: string | null;
  tenantId?: string | null;
  userId?: string | null;
  status: RepairKnowledgeStatus;
  source: RepairKnowledgeSource;
  helpfulCount: number;
  notHelpfulCount: number;
  createdAt: string;
}

export interface KnowledgeForJobResponse {
  checklist: FaultDiagnosis | null;
  notes: RepairNote[];
  suggestedFaultTypes: FaultType[] | null;
  jobDetails: {
    brand: string;
    model: string;
    problem: string;
    phoneModelId: string | null;
    faultTypeId: string | null;
  };
}

/**
 * Get full knowledge context for a job card
 */
export async function getKnowledgeForJob(jobCardId: string): Promise<KnowledgeForJobResponse> {
  const response = await authenticatedFetch(`/mobileshop/knowledge/job/${jobCardId}`);
  if (!response.ok) {
    const error = await extractData(response);
    throw new Error((error as any).message || "Failed to fetch knowledge for job");
  }
  return extractData(response);
}

/**
 * List all standard fault types
 */
export async function listFaultTypes(): Promise<FaultType[]> {
  const response = await authenticatedFetch(`/mobileshop/knowledge/fault-types`);
  if (!response.ok) {
    const error = await extractData(response);
    throw new Error((error as any).message || "Failed to fetch fault types");
  }
  return extractData(response);
}

/**
 * Vote on a repair note
 */
export async function voteOnNote(noteId: string, vote: "helpful" | "notHelpful"): Promise<void> {
  const response = await authenticatedFetch(`/mobileshop/knowledge/notes/${noteId}/vote`, {
    method: "POST",
    body: JSON.stringify({ vote }),
  });
  if (!response.ok) {
    const error = await extractData(response);
    throw new Error((error as any).message || "Failed to register vote");
  }
}

/**
 * Submit a new repair note
 */
export async function submitRepairNote(data: {
  phoneModelId?: string | null;
  faultTypeId: string;
  content: string;
  videoUrl?: string;
  source?: RepairKnowledgeSource;
}): Promise<RepairNote> {
  const response = await authenticatedFetch(`/mobileshop/knowledge/notes`, {
    method: "POST",
    body: JSON.stringify(data),
  });
  if (!response.ok) {
    const error = await extractData(response);
    throw new Error((error as any).message || "Failed to submit repair note");
  }
  return extractData(response);
}

/**
 * Moderate a note (Admin/Owner only)
 */
export async function moderateNote(noteId: string, status: RepairKnowledgeStatus): Promise<RepairNote> {
  const response = await authenticatedFetch(`/mobileshop/knowledge/notes/${noteId}/moderate`, {
    method: "PATCH",
    body: JSON.stringify({ status }),
  });
  if (!response.ok) {
    const error = await extractData(response);
    throw new Error((error as any).message || "Failed to moderate note");
  }
  return extractData(response);
}

/**
 * Create or update a diagnostic checklist
 */
export async function upsertChecklist(data: {
  faultTypeId: string;
  description?: string;
  steps: { order: number; stepText: string }[];
}): Promise<FaultDiagnosis> {
  const response = await authenticatedFetch(`/mobileshop/knowledge/checklist`, {
    method: "POST",
    body: JSON.stringify(data),
  });
  if (!response.ok) {
    const error = await extractData(response);
    throw new Error((error as any).message || "Failed to save checklist");
  }
  return extractData(response);
}

/**
 * Get diagnostic checklist for a fault type
 */
export async function getChecklist(faultTypeId: string): Promise<FaultDiagnosis | null> {
  const response = await authenticatedFetch(`/mobileshop/knowledge/checklist/${faultTypeId}`);
  if (!response.ok) {
    const error = await extractData(response);
    throw new Error((error as any).message || "Failed to fetch checklist");
  }
  return extractData(response);
}

/**
 * Get repair notes for a model and/or fault type
 */
export async function getRepairNotes(params: {
  phoneModelId?: string;
  faultTypeId?: string;
}): Promise<RepairNote[]> {
  const query = new URLSearchParams();
  if (params.phoneModelId) query.set("phoneModelId", params.phoneModelId);
  if (params.faultTypeId) query.set("faultTypeId", params.faultTypeId);

  const response = await authenticatedFetch(`/mobileshop/knowledge/notes?${query.toString()}`);
  if (!response.ok) {
    const error = await extractData(response);
    throw new Error((error as any).message || "Failed to fetch repair notes");
  }
  return extractData(response);
}
